---
title: "CAG low and high for human and mouse bulk RNA-seq analysis"
author: "Jian Zhang (zhang.jian@yale.edu)"
date: "`r format(Sys.Date())`"
output: 
  html_notebook: 
    code_folding: hide
  github_document: default
---

Experiment (12 samples):

13_1, 13_2, 13_3  Mouse (H2B MEF Low)

14_1, 14_2, 14_3  Mouse (H2B MEF High)

15_1, 15_2, 15_3  Human (Hu2F Low)

16_1, 16_2, 16_3  Human (Hu2F High)

# For mouse
Using *Tophat2* and *Cuffdiff* to do reads mapping, gene quantification and DEGs identification. Run the scripts and pepline in Yale linux server (`/home/zj76/project/Qiao_12_samples_CAG/6_mouse_samples`):

`perl 1.run.tophat2.parallel.pl /home/zj76/project/Qiao_12_samples_CAG/6_mouse_samples/Sample_13_1_to_Sample_14_3`

`perl 2.run.cuffdiff.with.slurm.pl /home/zj76/project/database/mm10/Mus_musculus.GRCm38.96.gtf mouse_13,mouse_14`

`sbatch cuffdiff.sh.run.slurm.sh`

`perl 3.mapping.sta.pl`

`perl 4.HTSeq.counts.table.pl`;`perl 4.FPKM.table.from.cuffnorm.pl`

`perl 5.FPKM.fiter.pl`

`perl 6.FPKM.sta.pl gene.FPKM.filtered.table`

`perl 7.DEGs.from.cuffdiff.pl`

```{r global_options, warning=FALSE, message=FALSE}
library(ggplot2)
library(pheatmap)
library(tidyverse)
library(dplyr)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(GO.db)
library(clusterProfiler)
library(enrichplot)
library(fgsea)
library(DOSE)
library(RcisTarget)
library(DT)
library(reshape2)
library(visNetwork)
```

## Mapping statistics
```{r}
mapping = read.table("6_mouse_samples/mapping.sta.table",header=T, row.names = 1)
row.names(mapping) = c("mouse_CAG_low_13_1","mouse_CAG_low_13_2","mouse_CAG_low_13_3","mouse_CAG_hi_14_1","mouse_CAG_hi_14_2","mouse_CAG_hi_14_3")
mapping
```
## Samples correlation
```{r}
fpkm = read.table("6_mouse_samples/gene.FPKM.filtered.table", header=T, row.names = 1)
#fpkm = read_table("6_mouse_samples/gene.FPKM.filtered.table")
colnames(fpkm) = c("mouse_CAG_low_13_1","mouse_CAG_low_13_2","mouse_CAG_low_13_3","mouse_CAG_hi_14_1","mouse_CAG_hi_14_2","mouse_CAG_hi_14_3")
b=cor(fpkm,method="kendall") #method = c("pearson", "kendall", "spearman")
pheatmap(b)
```
## Genes correlation
```{r}
pheatmap(log10(fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=F,
         show_colnames=T,
         fontsize_row=1,
         fontsize_col=10)
```
## Differential expressed genes (DEGs)
```{r}
all_genes = read.table("6_mouse_samples/DEGs/mouse_13.VS.mouse_14.all.table", header=T, row.names = 1)
degs = read.table("6_mouse_samples/DEGs/mouse_13.VS.mouse_14.DEGs.table", header=T, row.names = 1)
datatable(degs,filter="top")
```
```{r}
barplot(table(degs$regulation),col=c("blue","red"))
```
#### There are 2219 up-regulated and 1915 down-regulated genes in samples of mouse_CAG_hi compared to mouse_CAG_low
```{r fig.width=7, fig.height=7}
normal=read.table("6_mouse_samples/DEGs/mouse_13.VS.mouse_14.normal.for.plot",header=T)
up=read.table("6_mouse_samples/DEGs/mouse_13.VS.mouse_14.Up.for.plot",header=T)
down=read.table("6_mouse_samples/DEGs/mouse_13.VS.mouse_14.Down.for.plot",header=T)
plot(normal,type="p",pch=20,col="gray",xlab="Fold change [log2(mouse_CAG_hi/mouse_CAG_low)]",ylab="-Log10(FDR)",xlim=c(-5,5),ylim=c(0,4))
points(up,pch=20,col="red")
points(down,pch=20,col="blue")
abline(h=-log10(0.05),col="green")
abline(v=-1,col="green")
abline(v=1,col="green")
```
## GO enrichment
### BP subontologies for all DEGs
```{r warning=FALSE, message = FALSE}
#sapply(degs,mode)
degs_ensemble_id = as.character(degs$gene_id)
degs_gene_name = as.character(degs$gene)
degs_enrich_go_bp = enrichGO(gene=degs_gene_name,
                             OrgDb="org.Mm.eg.db",
                             keyType = "SYMBOL",
                             ont = "BP",
                             pvalueCutoff  = 0.01,
                             pAdjustMethod = "BH",
                             qvalueCutoff = 0.05,
                             minGSSize = 5, 
                             maxGSSize = 2000)
dotplot(degs_enrich_go_bp, title="GO (BP) enrichment for DEGs",showCategory=10)
```
### MF subontologies for all DEGs
```{r warning=FALSE, message = FALSE}
degs_enrich_go_mf = enrichGO(gene=degs_gene_name,
                             OrgDb="org.Mm.eg.db",
                             keyType = "SYMBOL",
                             ont = "MF",
                             pvalueCutoff  = 0.01,
                             pAdjustMethod = "BH",
                             qvalueCutoff = 0.05,
                             minGSSize = 5, 
                             maxGSSize = 2000)
dotplot(degs_enrich_go_mf, title="GO (MF) enrichment for DEGs",showCategory=10)
```
### BP subontologies for up- and down-regulated DEGs 
```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=8}
# GO enrichment for DEGs
formula_res <- compareCluster(
  gene~regulation,
  data=degs,
  fun="enrichGO",
  OrgDb="org.Mm.eg.db",
  keyType = "SYMBOL",
  ont		   = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05
)
# Run GO enrichment test and merge terms that are close to each other to remove result redundancy
lineage1_ego <- simplify(
  formula_res,
  cutoff=0.5,
  by="p.adjust",
  select_fun=min
)
```
```{r}
datatable(lineage1_ego@compareClusterResult,filter="top")
```

```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=8}
# Plot both analysis results
dotplot(lineage1_ego, title="GO (BP) enrichment for up and down-regulated DEGs", showCategory=10)
```

### Gene expression in special GO terms
#### Up-regulated genes in mouse_CAG_hi related to "cell-cell signaling by wnt"
```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=15}
all_genes = transform(all_genes, gene_id=as.character(gene_id))
all_genes = transform(all_genes, gene=as.character(gene))
all_genes = data.frame(all_genes)
genes_in_Go_term = lineage1_ego@compareClusterResult[lineage1_ego@compareClusterResult$Description == "cell-cell signaling by wnt",]$geneID
genes_in_Go_term = as.character(genes_in_Go_term)
genes_in_Go_term = unlist(strsplit(genes_in_Go_term,"/"))
genes_ensemble_ID_in_Go_term = all_genes[all_genes$gene %in% genes_in_Go_term,]$gene_id
genes_fpkm = fpkm[genes_ensemble_ID_in_Go_term,]
row.names(genes_fpkm) = all_genes[all_genes$gene_id %in% genes_ensemble_ID_in_Go_term,]$gene
pheatmap(log10(genes_fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=T,
         show_colnames=T,
         fontsize_row=10,
         fontsize_col=12)
```

#### Up-regulated genes in mouse_CAG_hi related to "regulation of cell-cell adhesion"
```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=15}
genes_in_Go_term = lineage1_ego@compareClusterResult[lineage1_ego@compareClusterResult$Description == "regulation of cell-cell adhesion",]$geneID
genes_in_Go_term = as.character(genes_in_Go_term)
genes_in_Go_term = unlist(strsplit(genes_in_Go_term,"/"))
genes_ensemble_ID_in_Go_term = all_genes[all_genes$gene %in% genes_in_Go_term,]$gene_id
genes_fpkm = fpkm[genes_ensemble_ID_in_Go_term,]
row.names(genes_fpkm) = all_genes[all_genes$gene_id %in% genes_ensemble_ID_in_Go_term,]$gene
pheatmap(log10(genes_fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=T,
         show_colnames=T,
         fontsize_row=10,
         fontsize_col=12)
```
#### Up-regulated genes in mouse_CAG_hi related to "muscle cell proliferation"
```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=15}
genes_in_Go_term = lineage1_ego@compareClusterResult[lineage1_ego@compareClusterResult$Description == "muscle cell proliferation",]$geneID
genes_in_Go_term = as.character(genes_in_Go_term)
genes_in_Go_term = unlist(strsplit(genes_in_Go_term,"/"))
genes_ensemble_ID_in_Go_term = all_genes[all_genes$gene %in% genes_in_Go_term,]$gene_id
genes_fpkm = fpkm[genes_ensemble_ID_in_Go_term,]
row.names(genes_fpkm) = all_genes[all_genes$gene_id %in% genes_ensemble_ID_in_Go_term,]$gene
pheatmap(log10(genes_fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=T,
         show_colnames=T,
         fontsize_row=10,
         fontsize_col=12)
```

## KEGG enrichment
```{r}
# #Convert Gene Symbols to Entrez IDs in R
# # Run columns(org.Hs.eg.db) to see available identifiers that can be used in this package. There are actually a lot of things such as Ensembl IDs, Uniprot IDs, protein families and GO annotations:
# columns(org.Mm.eg.db)
# columns(org.Hs.eg.db)
# # Letâ€™s make a sample gene symbol list to work with and do the conversion using mapIds which required 4 arguments, the first is the object itself, the second is the list of identifiers (symbols in this case), the third is the identifier type we want to convert to, and the last is the type of identifier for the second argument:
# symbols <- c('AHNAK', 'BOD1L1', 'HSPB1', 'SMARCA4', 'TRIM28')
# mapIds(org.Hs.eg.db, symbols, 'ENTREZID', 'SYMBOL')
```

```{r warning=FALSE, message = FALSE}
degs_gene_SYMBOL_to_UNIPROT = mapIds(org.Mm.eg.db, degs_gene_name, 'UNIPROT', 'SYMBOL')
# degs_gene_SYMBOL_to_UNIPROT = select(org.Mm.eg.db,
#                                      keys = degs_gene_name,
#                                      columns = c("UNIPROT","SYMBOL"),
#                                      keytype = "SYMBOL")
degs_KEGG_enrichment <- enrichKEGG(unname(degs_gene_SYMBOL_to_UNIPROT),
           organism = "mmu",
           keyType = "uniprot",
           pvalueCutoff  = 0.01,
           pAdjustMethod = "BH",
           qvalueCutoff = 0.05,
           minGSSize = 5,
           maxGSSize = 2000)#universe=all.genes,use_internal_data=F
dotplot(degs_KEGG_enrichment, title="KEGG enrichment for DEGs",showCategory=10)
```

#### Genes enrich in "Parkinson disease"
```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=15}
genes_in_kegg = degs_KEGG_enrichment@result[degs_KEGG_enrichment@result$Description == "Parkinson disease",]$geneID
genes_in_kegg = as.character(genes_in_kegg)
genes_in_kegg = unlist(strsplit(genes_in_kegg,"/"))
genes_in_kegg_UNIPROT_to_ensemble_ID = mapIds(org.Mm.eg.db, genes_in_kegg, 'ENSEMBL', 'UNIPROT')
genes_fpkm = fpkm[unname(genes_in_kegg_UNIPROT_to_ensemble_ID),]
row.names(genes_fpkm) = all_genes[all_genes$gene_id %in% unname(genes_in_kegg_UNIPROT_to_ensemble_ID),]$gene
pheatmap(log10(genes_fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=T,
         show_colnames=T,
         fontsize_row=10,
         fontsize_col=12)
```

#### Genes enrich in "Huntington disease"
```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=15}
genes_in_kegg = degs_KEGG_enrichment@result[degs_KEGG_enrichment@result$Description == "Huntington disease",]$geneID
genes_in_kegg = as.character(genes_in_kegg)
genes_in_kegg = unlist(strsplit(genes_in_kegg,"/"))
genes_in_kegg_UNIPROT_to_ensemble_ID = mapIds(org.Mm.eg.db, genes_in_kegg, 'ENSEMBL', 'UNIPROT')
genes_fpkm = fpkm[unname(genes_in_kegg_UNIPROT_to_ensemble_ID),]
row.names(genes_fpkm) = all_genes[all_genes$gene_id %in% unname(genes_in_kegg_UNIPROT_to_ensemble_ID),]$gene
pheatmap(log10(genes_fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=T,
         show_colnames=T,
         fontsize_row=10,
         fontsize_col=12)
```

#### Genes enrich in "Alzheimer disease"
```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=15}
genes_in_kegg = degs_KEGG_enrichment@result[degs_KEGG_enrichment@result$Description == "Alzheimer disease",]$geneID
genes_in_kegg = as.character(genes_in_kegg)
genes_in_kegg = unlist(strsplit(genes_in_kegg,"/"))
genes_in_kegg_UNIPROT_to_ensemble_ID = mapIds(org.Mm.eg.db, genes_in_kegg, 'ENSEMBL', 'UNIPROT')
genes_fpkm = fpkm[unname(genes_in_kegg_UNIPROT_to_ensemble_ID),]
row.names(genes_fpkm) = all_genes[all_genes$gene_id %in% unname(genes_in_kegg_UNIPROT_to_ensemble_ID),]$gene
pheatmap(log10(genes_fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=T,
         show_colnames=T,
         fontsize_row=10,
         fontsize_col=12)
```




## [RcisTarget](https://bioconductor.org/packages/release/bioc/vignettes/RcisTarget/inst/doc/RcisTarget.html): Transcription factor binding motif enrichment

RcisTarget is an R-package to identify transcription factor (TF) binding motifs over-represented on a gene list.

RcisTarget is based on the methods previously implemented in [i-cisTarget](https://gbiomed.kuleuven.be/apps/lcb/i-cisTarget/) (web interface, region-based) and [iRegulon](http://iregulon.aertslab.org/) (Cytoscape plug-in, gene-based).

If you use RcisTarget in your research, please cite: Aibar et al. (2017) SCENIC: single-cell regulatory network inference and clustering. Nature Methods. doi: 10.1038/nmeth.4463

To analyze the gene list, RcisTarget needs two types of databases: 

- 1. Gene-motif rankings: which provides the rankings (~score) of all the genes for each motif. 

- 2. The annotation of motifs to transcription factors.

### TFBS motif enrichment of Up-regulated genes in mouse_CAG_hi
```{r arning=FALSE, message = FALSE}
# Load gene sets to analyze. e.g.:
geneSet1 <- as.character(degs[degs$regulation == "up",]$gene)
geneLists <- list("UpDEGs" = geneSet1)

# Select motif database to use (i.e. organism and distance around TSS)
# mouse:
data(motifAnnotations_mgi)
# human:
#data(motifAnnotations_hgnc)

motifRankings <- importRankings("D:/cisTarget_databases/mm9-tss-centered-10kb-7species.mc9nr.feather")

# Motif enrichment analysis:
motifEnrichmentTable_wGenes <- cisTarget(geneLists, motifRankings,
                               motifAnnot=motifAnnotations_mgi)
```
* Advanced use
   * 1. Calculate enrichment
  The first step to estimate the over-representation of each motif on the gene-set is to calculate the Area Under the Curve (AUC) for each pair of motif-geneSet. This is calculated based on the recovery curve of the gene-set on the motif ranking (genes ranked decreasingly by the score of motif in its proximity, as provided in the motifRanking database).
   * 2. Select significant motifs and/or annotate to TFs
  The selection of significant motifs is done based on the Normalized Enrichment Score (NES). The NES is calculated -for each motif- based on the AUC distribution of all the motifs for the gene-set [(x-mean)/sd]. Those motifs that pass the given threshold (3.0 by default) are considered significant.
   * 3. Identify the genes with the best enrichment for each Motif
  Since RcisTarget searches for enrichment of a motif within a gene list, finding a motif â€˜enrichedâ€™ does not imply that all the genes in the gene-list have a high score for the motif. In this way, the third step of the workflow is to identify which genes (of the gene-set) are highly ranked for each of the significant motifs.
```{r warning=FALSE, message = FALSE}
# 1. Calculate AUC
motifs_AUC <- calcAUC(geneLists, motifRankings)

# 2. Select significant motifs, add TF annotation & format as table
motifEnrichmentTable <- addMotifAnnotation(motifs_AUC, 
                                           # nesThreshold=3,
                                           # highlightTFs=list(UpDEGs="Srf"),
                                           motifAnnot=motifAnnotations_mgi
                                           )

# 3. Identify significant genes for each motif
# (i.e. genes from the gene set in the top of the ranking)
# Note: Method 'iCisTarget' instead of 'aprox' is more accurate, but slower
motifEnrichmentTable_wGenes <- addSignificantGenes(motifEnrichmentTable, 
                                                   geneSets=geneLists,
                                                   # nCores=1,
                                                   # method="aprox",
                                                   rankings=motifRankings
                                                   )
```

* Output

The final output of RcisTarget is a data.table containing the information about the motif enrichment and its annotation organized in the following fields:
   * geneSet: Name of the gene set
   * motif: ID of the motif
   * NES: Normalized enrichment score of the motif in the gene-set
   * AUC: Area Under the Curve (used to calculate the NES)
   * TFinDB: Indicates whether the highlightedTFs are included within the high confidence annotation (two asterisks) or low confidence annotation (one asterisk).
   * TF_highConf: Transcription factors annotated to the motif according to â€˜motifAnnot_highConfCatâ€™.
   * TF_lowConf: Transcription factors annotated to the motif according to â€˜motifAnnot_lowConfCatâ€™.
   * enrichedGenes: Genes that are highly ranked for the given motif.
   * nErnGenes: Number of genes highly ranked
   * rankAtMax: Ranking at the maximum enrichment, used to determine the number of enriched genes.
```{r warning=FALSE, message = FALSE}
motifEnrichmentTable_wGenes_wLogo <- addLogo(motifEnrichmentTable_wGenes)

resultsSubset <- motifEnrichmentTable_wGenes_wLogo[1:20,]

datatable(resultsSubset[,-c("enrichedGenes", "TF_lowConf"), with=FALSE], 
          escape = FALSE, # To show the logo
          filter="top", options=list(pageLength=5))
```

* TFs annotated to the enriched motifs

Note that the TFs are provided based on the motif annotation. They can be used as a guide to select relevant motifs or to prioritize some TFs, but the motif annotation does not imply that all the TFs appearing in the table regulate the gene list.
```{r}
anotatedTfs <- lapply(split(motifEnrichmentTable_wGenes$TF_highConf,
                            motifEnrichmentTable$geneSet),
                      function(x) {
                        genes <- gsub(" \\(.*\\). ", "; ", x, fixed=FALSE)
                        genesSplit <- unique(unlist(strsplit(genes, "; ")))
                        return(genesSplit)
                        })
                      
anotatedTfs$UpDEGs
```


* Building a network
```{r}
signifMotifNames <- motifEnrichmentTable$motif[1:3]

incidenceMatrix <- getSignificantGenes(geneLists$UpDEGs, 
                                       motifRankings,
                                       signifRankingNames=signifMotifNames,
                                       plotCurve=TRUE, maxRank=5000-20, 
                                       genesFormat="incidMatrix",
                                       method="aprox")$incidMatrix

edges <- melt(incidenceMatrix)
edges <- edges[which(edges[,3]==1),1:2]
colnames(edges) <- c("from","to")
```

* Network Visualization
```{r fig.width=10, fig.height=10}
motifs <- unique(as.character(edges[,1]))
genes <- unique(as.character(edges[,2]))
nodes <- data.frame(id=c(motifs, genes),   
      label=c(motifs, genes),    
      title=c(motifs, genes), # tooltip 
      shape=c(rep("diamond", length(motifs)), rep("elypse", length(genes))),
      color=c(rep("purple", length(motifs)), rep("skyblue", length(genes))))
visNetwork(nodes, edges) %>% visOptions(highlightNearest = TRUE, 
                                        nodesIdSelection = TRUE)
```


### TFBS motif enrichment of Down-regulated genes in mouse_CAG_hi
```{r arning=FALSE, message = FALSE}
# Load gene sets to analyze. e.g.:
geneSet1 <- as.character(degs[degs$regulation == "down",]$gene)
geneLists <- list("DownDEGs" = geneSet1)

# Select motif database to use (i.e. organism and distance around TSS)
# mouse:
data(motifAnnotations_mgi)
# human:
#data(motifAnnotations_hgnc)

motifRankings <- importRankings("D:/cisTarget_databases/mm9-tss-centered-10kb-7species.mc9nr.feather")

# Motif enrichment analysis:
motifEnrichmentTable_wGenes <- cisTarget(geneLists, 
                                         motifRankings,
                                         motifAnnot=motifAnnotations_mgi)
```
```{r warning=FALSE, message = FALSE}
# 1. Calculate AUC
motifs_AUC <- calcAUC(geneLists, motifRankings)

# 2. Select significant motifs, add TF annotation & format as table
motifEnrichmentTable <- addMotifAnnotation(motifs_AUC, 
                                           # nesThreshold=3,
                                           # highlightTFs=list(UpDEGs="Srf"),
                                           motifAnnot=motifAnnotations_mgi
                                           )

# 3. Identify significant genes for each motif
# (i.e. genes from the gene set in the top of the ranking)
# Note: Method 'iCisTarget' instead of 'aprox' is more accurate, but slower
motifEnrichmentTable_wGenes <- addSignificantGenes(motifEnrichmentTable, 
                                                   geneSets=geneLists,
                                                   # nCores=1,
                                                   # method="aprox",
                                                   rankings=motifRankings
                                                   )
```
```{r warning=FALSE, message = FALSE}
motifEnrichmentTable_wGenes_wLogo <- addLogo(motifEnrichmentTable_wGenes)

resultsSubset <- motifEnrichmentTable_wGenes_wLogo[1:20,]

datatable(resultsSubset[,-c("enrichedGenes", "TF_lowConf"), with=FALSE], 
          escape = FALSE, # To show the logo
          filter="top", options=list(pageLength=5))
```
* TFs annotated to the enriched motifs
```{r}
anotatedTfs <- lapply(split(motifEnrichmentTable_wGenes$TF_highConf,
                            motifEnrichmentTable$geneSet),
                      function(x) {
                        genes <- gsub(" \\(.*\\). ", "; ", x, fixed=FALSE)
                        genesSplit <- unique(unlist(strsplit(genes, "; ")))
                        return(genesSplit)
                        })
                      
anotatedTfs$DownDEGs
```

* Building a network
```{r}
signifMotifNames <- motifEnrichmentTable$motif[1:3]

incidenceMatrix <- getSignificantGenes(geneLists$DownDEGs, 
                                       motifRankings,
                                       signifRankingNames=signifMotifNames,
                                       plotCurve=TRUE, maxRank=5000-20, 
                                       genesFormat="incidMatrix",
                                       method="aprox")$incidMatrix

edges <- melt(incidenceMatrix)
edges <- edges[which(edges[,3]==1),1:2]
colnames(edges) <- c("from","to")
```

* Network Visualization
```{r  fig.width=10, fig.height=10}
motifs <- unique(as.character(edges[,1]))
genes <- unique(as.character(edges[,2]))
nodes <- data.frame(id=c(motifs, genes),   
      label=c(motifs, genes),    
      title=c(motifs, genes), # tooltip 
      shape=c(rep("diamond", length(motifs)), rep("elypse", length(genes))),
      color=c(rep("purple", length(motifs)), rep("skyblue", length(genes))))
visNetwork(nodes, edges) %>% visOptions(highlightNearest = TRUE, 
                                        nodesIdSelection = TRUE)
```
