---
title: "CAG low and high for human and mouse bulk RNA-seq analysis"
author: "Jian Zhang (zhang.jian@yale.edu)"
date: "`r format(Sys.Date())`"
output: 
  html_notebook: 
    code_folding: hide
  github_document: default
---

# For human
Run the same scripts and pepline with mouse in Yale linux server (`/home/zj76/project/Qiao_12_samples_CAG/6_human_samples`):

`perl 1.run.tophat2.parallel.pl /home/zj76/project/Qiao_12_samples_CAG/6_human_samples/Sample_15_1_to_Sample_16_3`

`perl 2.run.cuffdiff.with.slurm.pl /home/zj76/project/database/hg38/Homo_sapiens.GRCh38.96.gtf human_15,human_16`

`sbatch cuffdiff.sh.run.slurm.sh`

`perl 3.mapping.sta.pl`

`perl 4.HTSeq.counts.table.pl`;`perl 4.FPKM.table.from.cuffnorm.pl`

`perl 5.FPKM.fiter.pl`

`perl 6.FPKM.sta.pl gene.FPKM.filtered.table`

`perl 7.DEGs.from.cuffdiff.pl`

```{r global_options, warning=FALSE, message=FALSE}
library(ggplot2)
library(pheatmap)
library(tidyverse)
library(dplyr)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(GO.db)
library(clusterProfiler)
library(enrichplot)
library(fgsea)
library(DOSE)
library(RcisTarget)
library(DT)
library(reshape2)
library(visNetwork)
```

## Mapping statistics
```{r}
mapping = read.table("6_human_samples/mapping.sta.table",header=T, row.names = 1)
row.names(mapping) = c("human_CAG_low_15_1","human_CAG_low_15_2","human_CAG_low_15_3","human_CAG_hi_16_1","human_CAG_hi_16_2","human_CAG_hi_16_3")
mapping
```

## Samples correlation
```{r}
fpkm = read.table("6_human_samples/gene.FPKM.filtered.table", header=T, row.names = 1)
colnames(fpkm) = c("human_CAG_low_15_1","human_CAG_low_15_2","human_CAG_low_15_3","human_CAG_hi_16_1","human_CAG_hi_16_2","human_CAG_hi_16_3")
b=cor(fpkm,method="kendall") #method = c("pearson", "kendall", "spearman")
pheatmap(b)
```
## Genes correlation
```{r}
pheatmap(log10(fpkm+1),
         cluster_rows=T,
         cluster_cols=F,
         show_rownames=F,
         show_colnames=T,
         fontsize_row=1,
         fontsize_col=10)
```
## Differential expressed genes (DEGs)
```{r}
all_genes = read.table("6_human_samples/DEGs/human_15.VS.human_16.all.table", header=T, row.names = 1)
degs = read.table("6_human_samples/DEGs/human_15.VS.human_16.DEGs.table", header=T, row.names = 1)
datatable(degs,filter="top")
```
```{r}
barplot(table(degs$regulation),col=c("blue","red"))
```

#### There are 144 up-regulated and 41 down-regulated genes in samples of human_CAG_hi compared to human_CAG_low
```{r fig.width=7, fig.height=7}
normal=read.table("6_human_samples/DEGs/human_15.VS.human_16.normal.for.plot",header=T)
up=read.table("6_human_samples/DEGs/human_15.VS.human_16.Up.for.plot",header=T)
down=read.table("6_human_samples/DEGs/human_15.VS.human_16.Down.for.plot",header=T)
plot(normal,type="p",pch=20,col="gray",xlab="Fold change [log2(human_CAG_hi/human_CAG_low)]",ylab="-Log10(FDR)",xlim=c(-5,5),ylim=c(0,4))
points(up,pch=20,col="red")
points(down,pch=20,col="blue")
abline(h=-log10(0.05),col="green")
abline(v=-1,col="green")
abline(v=1,col="green")
```



## GO enrichment
### BP subontologies for all DEGs
```{r warning=FALSE, message = FALSE}
degs_gene_name = as.character(degs$gene)
degs_enrich_go_bp = enrichGO(gene=degs_gene_name,
                             OrgDb="org.Hs.eg.db",
                             keyType = "SYMBOL",
                             ont = "BP",
                             pvalueCutoff  = 0.01,
                             pAdjustMethod = "BH",
                             qvalueCutoff = 0.05,
                             minGSSize = 5, 
                             maxGSSize = 2000)
dotplot(degs_enrich_go_bp, title="GO (BP) enrichment for DEGs",showCategory=10)
```

### MF subontologies for all DEGs
```{r warning=FALSE, message = FALSE}
degs_enrich_go_mf = enrichGO(gene=degs_gene_name,
                             OrgDb="org.Hs.eg.db",
                             keyType = "SYMBOL",
                             ont = "MF",
                             pvalueCutoff  = 0.01,
                             pAdjustMethod = "BH",
                             qvalueCutoff = 0.05,
                             minGSSize = 5, 
                             maxGSSize = 2000)
dotplot(degs_enrich_go_mf, title="GO (MF) enrichment for DEGs",showCategory=10)
```

```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=6}
# GO enrichment for DEGs
formula_res <- compareCluster(
  gene~regulation,
  data=degs,
  fun="enrichGO",
  OrgDb="org.Hs.eg.db",
  keyType = "SYMBOL",
  ont		   = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05
)
# Run GO enrichment test and merge terms that are close to each other to remove result redundancy
lineage1_ego <- simplify(
  formula_res,
  cutoff=0.5,
  by="p.adjust",
  select_fun=min
)

```
```{r}
datatable(lineage1_ego@compareClusterResult,filter="top")
```

```{r warning=FALSE, message = FALSE, fig.width=10, fig.height=6}
# Plot both analysis results
dotplot(lineage1_ego, title="GO (BP) enrichment for up and down-regulated DEGs", showCategory=6)
```


## KEGG enrichment
```{r warning=FALSE, message = FALSE}
degs_gene_SYMBOL_to_UNIPROT = mapIds(org.Hs.eg.db, degs_gene_name, 'UNIPROT', 'SYMBOL')
p <- enrichKEGG(unname(degs_gene_SYMBOL_to_UNIPROT),
           organism = "hsa",
           keyType = "uniprot",
           # pvalueCutoff  = 0.01,
           # pAdjustMethod = "BH",
           # qvalueCutoff = 0.05,
           minGSSize = 5,
           maxGSSize = 2000)#universe=all.genes,use_internal_data=F
dotplot(p, title="KEGG enrichment for DEGs",showCategory=10)
```

